useradmin/md5-lib.pl

```diff
@@ -205,14 +205,39 @@ sub check_sha512
 }
 
 # encrypt_sha512(password, [salt])
-# Hashes a password, possibly with the give salt, with SHA512
+# Hashes a password, possibly with the given salt, with SHA512
 sub encrypt_sha512
 {
 my ($passwd, $salt) = @_;
 $salt ||= '$6$'.substr(time(), -8).'$';
 return crypt($passwd, $salt);
 }
 
+# unix_crypt_supports_yescrypt()
+# Returns 1 if the built-in crypt() function can already do yescrypt
+sub unix_crypt_supports_yescrypt
+{
+my $hash = '$y$j9T$waHytoaqP/CEnKFroGn0S/$fxd5mVc2mBPUc3vv.cpqDckpwrWTyIm2iD4JfnVBi26';
+my $newhash = eval { crypt('test', $hash) };
+return $newhash eq $hash;
+}
+
+# check_yescrypt()
+# Returns undef if yescrypt hashing is supported, or an error message if not
+sub check_yescrypt
+{
+return &unix_crypt_supports_yescrypt() ? undef : 'Crypt::NaCl::Sodium';
+}
+
+# encrypt_yescrypt(password, [salt])
+# Hashes a password, possibly with the given salt, with yescrypt
+sub encrypt_yescrypt
+{
+my ($passwd, $salt) = @_;
+$salt ||= &substitute_pattern('$y$j9T$[A-Z]{4}.[a-zA-Z0-9]{16}.$[a-zA-Z0-9]{14}.[a-zA-Z0-9]{7}/[a-zA-Z0-9]{15}/[a-zA-Z0-9]{4}');
+return crypt($passwd, $salt);
```

Comment:
Shouldn't you call `Crypt::NaCl::Sodium` if installed?

Reply:
No, cause system supports it.

Reply:
The most important part here is to pass salt in expected format, otherwize it just won't work. I'm glad I have found another application to my `substitute_pattern` sub. 🙂

Reply:
Why check if `Crypt::NaCl::Sodium` is installed above though?

Reply:
> Why check if `Crypt::NaCl::Sodium` is installed above though?

It does the same as your `check_sha512` - if system doesn't support it then it will provide a link to install it from CPAN.

Reply:
I have checked once more, and it looks that `usave_edigestsha512` and as a result `usave_edigestyescrypt` won't suggest installing from CPAN disregard, and will only suggest to change encryption method. Is this intended that you didn't include `Crypt::SHA` at the first time to be installable from CPAN?

Reply:
Hmm, I don't know why I did that with SHA512 either! It should be implemented more like `encrypt_md5`

Reply:
> Hmm, I don't know why I did that with SHA512 either! It should be implemented more like encrypt_md5

I have made a [patch](https://github.com/webmin/webmin/pull/1652/commits/dae7f8a397d31902c6a50421cbe4297e993ce4ac) for this to improve error message handling. Although, I assume we cannot (should not) install any Perl modules in case of `crypt` cannot handle it .. as I don't think it will be possible to authenticate even if we will be able to get the right hash.

Reply:
`crypt` doesn't depend on the perl modules though.

Take a look at how `encrypt_md5` works - if crypt doesn't support MD5, it falls back to calling a perl module.

Reply:
This is because, I assume MD5 is supported by all systems. But, I just think that if we install Perl `Crypt::NaCl::Sodium` module, and generate `yescrypt` hash and then put this to the shadow file, user still won't be able to authenticate! So I would expect, if system `crypt` doesn't happen to work then it won't work system-wide. This is why installing those isn't an option. Also, building from source is required, which isn't a good thing for a production system.

Reply:
Whether `yescrypt` is actually supported by PAM and the rest of the system is a separate question though. It seems quite possible that even if the older `crypt` libc method doesn't support this format, it might still be supported when logging in.

Reply:
> Whether yescrypt is actually supported by PAM and the rest of the system is a separate question though.

I was trying to say that this is frutial point.

> It seems quite possible that even if the older crypt libc method doesn't support this format, it might still be supported when logging in.

Certainly not. It can be easily tested by editing `/etc/shadow` file on Debian 10 and re-placing a hash for the _root_ user with _yescrypt_ from Ubuntu 22.04, and it won't work.

~~~
May 29 03:28:05 debug-debian10-gpl.local sshd[16488]: pam_unix(sshd:auth): authentication failure; logname= uid=0 euid=0 tty=ssh ruser= rhost=10.211.55.2  user=root
May 29 03:28:07 debug-debian10-gpl.local sshd[16488]: Failed password for root from 10.211.55.2 port 49967 ssh2
May 29 03:28:07 debug-debian10-gpl.local sshd[16488]: Connection closed by authenticating user root 10.211.55.2 port 49967 [preauth]
~~~

Reply:
Understood, but do we know for sure that all systems on which `crypt` supports `yescrypt` also support it in `/etc/shadow` ?  That would depend on how the PAM library interprets the shadow file entries.

Reply:
Yes, this should always be the case, as we read it from PAM common password config itself.